app/routes/questions/cache.tsx

```diff
@@ -1,17 +1,74 @@
-import {useLoaderData, Form} from '@remix-run/react'
-import {loadCache, cleanCache} from '~/server-utils/kv-cache'
+import {ActionArgs, json} from '@remix-run/cloudflare'
+import {useLoaderData, useActionData, useTransition, Form} from '@remix-run/react'
+import {useEffect, useState} from 'react'
+import {loadCacheKeys, loadCacheValue, cleanCache} from '~/server-utils/kv-cache'
 
-export const loader = async () => await loadCache()
+enum Actions {
+  delete = 'delete',
+  reload = 'reload',
+  loadCache = 'loadCache',
+}
+
+export const loader = async () => await loadCacheKeys()
 
-export const action = async () => await cleanCache()
+export const action = async ({request}: ActionArgs) => {
+  const data = Array.from(await request.formData()) as [Actions, string][]
+  if (data.length !== 1) {
+    return json(
+      {error: `Expected only 1 formData key-value pair, got ${JSON.stringify(data)}`},
+      {status: 400}
+    )
+  }
+  const [actionKey, cacheKey] = data[0]
+  if (actionKey === Actions.delete) {
+    return cleanCache(cacheKey)
+  } else if (actionKey === Actions.loadCache) {
+    return json({cacheKey, cacheValue: await loadCacheValue(cacheKey)})
+  } else if (actionKey === Actions.reload) {
+    return null
+  } else {
+    return json({error: `Unknown action ${actionKey}`}, {status: 400})
+  }
+}
 
 export default function Cache() {
-  const data = useLoaderData()
+  const keys = useLoaderData<typeof loader>()
+  const actionData = useActionData<typeof action>()
+  const transition = useTransition()
+  // @ts-expect-error inferred type kinda looks OK, but TS is unhappy anyway
+  const {error, cacheKey, cacheValue} = actionData ?? {}
+  const [cacheValues, setCacheValues] = useState(() =>
+    Object.fromEntries(keys.map((k) => [k, null]))
+  )
+  useEffect(() => {
+    if (cacheKey) {
+      setCacheValues((curr) => ({
+        ...curr,
+        [cacheKey]: cacheValue,
+      }))
+    }
+  }, [cacheKey, cacheValue])
 
   return (
-    <Form method="post">
-      <button>Clean cache</button>
-      <pre>{JSON.stringify(data, null, 2)}</pre>
+    <Form method="post" className="cache">
+      <h1>Cached data</h1>
+      <button name={Actions.delete}>Clean cache</button>
+      <button name={Actions.reload}>Reload keys</button>
+      {transition.type !== 'idle' && transition.type}
+      {error && <div className="error">{error}</div>}
+      <h2>Keys:</h2>
+      <ul>
+        {keys.length === 0 && <i>(the cache is empty)</i>}
+        {keys.map((key) => (
+          <li key={key}>
+            {key}
+            <button name={Actions.loadCache} value={key}>
+              {cacheValues[key] ? 'Reload' : 'Show'} value
+            </button>
+            {cacheValues[key] && <pre>{cacheValues[key]}</pre>}
```

Comment:
Could you do `<pre>{JSON.stringify(cacheValues[key], null, 2)}</pre>`?

Reply:
it's a string from KV cache, not parsed to JSON here .. parsing and stringifying to new format might slow down future bug investigation if we somehow make a really stupid value

it's possible to copy&paste to external editor though.. but feel free to improve by adding commits if you can handle the edge case of invalid or too large JSON